packages/marketprice/src/alarms/mod.rs

```diff
@@ -335,74 +450,185 @@ pub mod tests {
+            .unwrap();
+        alarms
+            .add_alarm_below(
+                subscriber1.clone(),
+                price::total_of::<Atom>(1.into()).is::<Base>(2.into()),
+            )
+            .unwrap();
+
+        alarms.ensure_no_in_delivery().unwrap();
+
+        alarms
+            .add_alarm_below(subscriber2.clone(), Price::<Atom, Base>::identity())
+            .unwrap();
+        alarms
+            .add_alarm_below(
+                subscriber2.clone(),
+                price::total_of::<Atom>(1.into()).is::<Base>(2.into()),
+            )
+            .unwrap();
+
+        alarms.ensure_no_in_delivery().unwrap();
+
+        alarms.out_for_delivery(subscriber1).unwrap();
+        alarms.out_for_delivery(subscriber2).unwrap();
+
+        assert!(matches!(
+            alarms.ensure_no_in_delivery().unwrap_err(),
+            AlarmError::NonEmptyAlarmsInDeliveryQueue(_)
+        ));
+
+        alarms.last_delivered().unwrap();
+
+        assert!(matches!(
+            alarms.ensure_no_in_delivery().unwrap_err(),
+            AlarmError::NonEmptyAlarmsInDeliveryQueue(_)
+        ));
+
+        alarms.last_delivered().unwrap();
+
+        alarms.ensure_no_in_delivery().unwrap();
+    }
+
+    #[test]
+    fn test_failed() {
+        let mut storage = MockStorage::new();
+        let mut alarms = alarms(&mut storage);
+
+        let subscriber1 = Addr::unchecked("subscriber1");
+        let subscriber2 = Addr::unchecked("subscriber2");
+
+        alarms.ensure_no_in_delivery().unwrap();
+
+        alarms
+            .add_alarm_below(subscriber1.clone(), Price::<Atom, Base>::identity())
+            .unwrap();
+        alarms
+            .add_alarm_below(
+                subscriber1.clone(),
+                price::total_of::<Atom>(1.into()).is::<Base>(2.into()),
+            )
+            .unwrap();
+
+        alarms.ensure_no_in_delivery().unwrap();
+
+        alarms
+            .add_alarm_below(subscriber2.clone(), Price::<Atom, Base>::identity())
+            .unwrap();
+        alarms
+            .add_alarm_below(
+                subscriber2.clone(),
+                price::total_of::<Atom>(1.into()).is::<Base>(2.into()),
+            )
+            .unwrap();
+
+        alarms.ensure_no_in_delivery().unwrap();
+
+        alarms.out_for_delivery(subscriber1).unwrap();
+        alarms.out_for_delivery(subscriber2).unwrap();
+
+        assert!(matches!(
+            alarms.ensure_no_in_delivery().unwrap_err(),
+            AlarmError::NonEmptyAlarmsInDeliveryQueue(_)
+        ));
+
+        alarms.last_delivered().unwrap();
+
+        assert!(matches!(
+            alarms.ensure_no_in_delivery().unwrap_err(),
+            AlarmError::NonEmptyAlarmsInDeliveryQueue(_)
+        ));
+
+        alarms.last_failed().unwrap();
+
+        assert!(matches!(
+            alarms.ensure_no_in_delivery().unwrap_err(),
+            AlarmError::NonEmptyAlarmsInDeliveryQueue(_)
+        ));
```

Comment:
am I mistaken or do we have a bug at this point? As far as I understand we should have nothing in delivery past `last_failed` - because it would pop the last item in delivery

Reply:
Good catch! Thanks.

Reply:
Good news is that CI caught it too.